python/ray/serve/drivers.py

```diff
@@ -80,16 +82,51 @@ async def __call__(self, request: starlette.requests.Request):
 
 
 @serve.deployment(route_prefix="/")
-class DAGDriver(SimpleSchemaIngress):
+class DAGDriver:
     def __init__(
         self,
-        dag_handle: RayServeDAGHandle,
-        *,
+        dags: Union[RayServeDAGHandle, Dict[str, RayServeDAGHandle]],
         http_adapter: Optional[Union[str, Callable]] = None,
     ):
-        self.dag_handle = dag_handle
-        super().__init__(http_adapter)
+        install_serve_encoders_to_fastapi()
+        http_adapter = load_http_adapter(http_adapter)
+        self.app = FastAPI()
+
+        if isinstance(dags, dict):
+            self.dags = dags
+            for route in dags:
+
+                @self.app.get(f"{route}")
+                @self.app.post(f"{route}")
+                async def handle_request(
+                    request: starlette.requests.Request, inp=Depends(http_adapter)
+                ):
+                    return await self.multi_dag_predict(request.url.path, inp)
```

Comment:
```suggestion
            for route, handle in dags.items():

                @self.app.get(f"{route}")
                @self.app.post(f"{route}")
                async def handle_request(
                    request: starlette.requests.Request, inp=Depends(http_adapter)
                ):
                    return await self.handle.remote(inp)
```

Now we don't need the special mutli_dag predict

We can also generalize the single DAG case into a wildcard route "/{path}"

Reply:
Hi Simon, 
i give a try, I think the handle variable life cycle it outlived after for loop.  All the endpoint will use the **last** handle 

For single DAG case into a wildcard route "/{path}", the change you suggest will receive all requests. Which means, we will not return 404 from FlaskAPI side (we have to do it by ourselves) when user send http://127.0.0.1/not_exist for single DAG.

Reply:
> All the endpoint will use the last handle
You can use functools.partial to properly perform the bind in arguments and closure: https://docs.python-guide.org/writing/gotchas/#late-binding-closures


> we will not return 404 from FlaskAPI side (we have to do it by ourselves) when user send http://127.0.0.1/not_exist for single DAG.

But for a single DAG we never return 404 because it's route_prefix enforced to "/{path}" already. User don't specify the route_prefix for single DAG case; if they want, they should do {"prefix": DAG} and pass the dict instead.

Reply:
Wow, works like a charm for functools.partial. Updated